x/metoken/keeper/reserves.go

```diff
@@ -0,0 +1,81 @@
+package keeper
+
+import (
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/umee-network/umee/v5/x/metoken"
+)
+
+// RebalanceReserves checks if the portion of reserves is below the desired and transfer the missing amount from
+// x/leverage to x/metoken reserves, or vice versa.
+func (k Keeper) RebalanceReserves() error {
+	rebalancingTime, err := k.getNextRebalancingTime()
+	if err != nil {
+		return err
+	}
+
+	if k.ctx.BlockTime().After(rebalancingTime) {
+		indexes := k.GetAllRegisteredIndexes()
+		for _, index := range indexes {
+			balances, err := k.IndexBalances(index.Denom)
+			if err != nil {
+				return err
+			}
+
+			// if no meToken were minted, there is nothing to check
+			if balances.MetokenSupply.IsPositive() {
+				updatedBalances := make([]metoken.AssetBalance, 0)
+				for _, balance := range balances.AssetBalances {
+					if balance.AvailableSupply().IsPositive() {
+						i, assetSettings := index.AcceptedAsset(balance.Denom)
+						if i < 0 {
+							return sdkerrors.ErrNotFound.Wrapf("asset %s not found in the index", balance.Denom)
+						}
+
+						// Calculate the desired reserves amount
+						desiredReserves := assetSettings.ReservePortion.MulInt(balance.AvailableSupply()).TruncateInt()
+						// In case the x/metoken module has fewer reserves than required,
+						// transfer the missing amount from x/leverage to x/metoken
+						if desiredReserves.GT(balance.Reserved) {
+							missingReserves := desiredReserves.Sub(balance.Reserved)
+							tokensWithdrawn, err := k.withdrawFromLeverage(sdk.NewCoin(balance.Denom, missingReserves))
+							if err != nil {
+								return err
```

Comment:
We cannot continue if we half-executed the `Withdraw` function from `x/leverage`, because the state will be compromised. That's why before withdrawing and supplying I check the `x/leverage` conditions. However, while checking the conditions some errors can be returned as well and in these particular cases we may not want to halt the chain. I will add an extra layer for errors returned from `x/leverage` to check if the error is recoverable or not. If it is, I will just log the error, but if not we can't avoid the panic.

Reply:
cc @toteki ^

Reply:
This should be easier if we add #2170